### PR TITLE
Replace sanitizer coverage with clang source-code based coverage

### PR DESCRIPTION
**Pull request by Shadoom7**
This branch replaced sanitizer coverage with clang-based coverage.
One issue is that some trials will end early without being preempted. Don't know is it inherited from the master or is it due to this patch.

### CODE REVIEW

**Comment by jonathanmetzman** [comment]
What is the difference between this branch and #565 they seem to have a lot of duplicated code.

**Comment by Shadoom7** [comment]
> What is the difference between this branch and #565 they seem to have a lot of duplicated code.

#565 is coded on top of this branch. Should first merge this branch and ignore #565, I will change #565 into draft for now.

File: experiment/test_run_coverage.py [hunk 1]
```diff
@@ -37,64 +37,67 @@ def _make_crashes_dir(parent_path):
...
 
 
-def _make_sancov_dir(parent_path):
-    """Makes a sancov dir in |parent_path| and returns it."""
-    sancov_dir = os.path.join(str(parent_path), 'sancov')
-    os.mkdir(sancov_dir)
-    return sancov_dir
+def _make_profraw_dir(parent_path):
+    """Makes a profraw dir in |parent_path| and returns it."""
+    profraw_dir = os.path.join(str(parent_path), 'profraw')
```
Comment (inferno-chromium) [inline experiment/test_run_coverage.py:42]: coverage_dir

File: experiment/test_run_coverage.py [hunk 2]
```diff
@@ -37,64 +37,67 @@ def _make_crashes_dir(parent_path):
...
-    pattern += '*.sancov'
+    pattern += '*.profraw'
     assert glob.glob(pattern)
 
 
 class TestIntegrationRunCoverage:
     """Integration tests for run_coverage.py"""
 
-    COVERAGE_BINARY_PATH = os.path.join(TEST_DATA_PATH, 'fuzz-target')
+    COVERAGE_BINARY_PATH = os.path.join(TEST_DATA_PATH, 'fuzz-target-clang')
```
Comment (inferno-chromium) [inline experiment/test_run_coverage.py:59]: -clang is a weird suffix.
-clang-cov makes sense.

File: fuzzers/coverage/builder.Dockerfile [hunk 3]
```diff
@@ -15,15 +15,12 @@
...
-    bash build.sh && \
-    cp libFuzzer.a /usr/lib
\ No newline at end of file
+    cd /llvm-project/ && \
+    git checkout d8981ce5b9f8caa567613b2bf5aa3095e0156130 && \
+    cd compiler-rt/lib/fuzzer && \
+    (for f in *.cpp; do \
+      clang++ -stdlib=libc++ -fPIC -O2 -std=c++11 $f -c & \
+    done && wait) && \
+    ar r /usr/lib/libFuzzer.a *.o
```
Comment (inferno-chromium) [inline fuzzers/coverage/builder.Dockerfile:26]: new line after this.

File: fuzzers/coverage/fuzzer.py [hunk 4]
```diff
@@ -11,7 +11,7 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-"""Integration code for coverage builds."""
+"""Integration code for source-based coverage builds."""
```
Comment (inferno-chromium) [inline fuzzers/coverage/fuzzer.py:14]: clang source-based

File: fuzzers/coverage/builder.Dockerfile [hunk 5]
```diff
@@ -15,15 +15,12 @@
 ARG parent_image=gcr.io/fuzzbench/base-builder
 FROM $parent_image
 
-# Minor modification to dump_coverage on a successful inner merge step.
-COPY patch.diff /
-
-# Use old libFuzzer code that supports -dump_coverage=1/trace-pc-guard coverage.
-# This is the last libFuzzer version before support was removed.
+# Use a libFuzzer version that supports source-based coverage
```
Comment (inferno-chromium) [inline fuzzers/coverage/builder.Dockerfile:18]: End comment with period.

**Comment by inferno-chromium** [comment]
You also need to update the text in report generation, it should be 'Region coverage' or 'Regions covered' or ??
```
analysis/plotting.py:        axes.set(ylabel='Edge coverage')
analysis/plotting.py:        axes.set(ylabel='Reached edge coverage')
analysis/plotting.py:        axes.set(xlabel='Edge coverage')
analysis/plotting.py:        axes.set(ylabel='Reached edge coverage')
```

File: experiment/measurer.py [hunk 6]
```diff
@@ -355,42 +360,48 @@ def run_cov_new_units(self):
...
-    def get_current_pcs(self) -> Set[str]:
-        """Get the current pcs covered by a fuzzer."""
-        with open(self.covered_pcs_filename) as file_handle:
-            current_pcs = set(
-                pc.strip() for pc in file_handle.readlines() if pc.strip())
-        return current_pcs
+    def get_current_coverage(self) -> int:
+        """Get the current number of lines covered"""
+        with open(self.summary_file) as summary:
+            coverage_info = json.loads(summary.readlines()[-1])
```
[the remark this answers is not in the document]
Comment (Shadoom7) [reply experiment/measurer.py:372]: For bloaty, the first line would be a warning info and the second line is the json we want. Only did this way to skip that warning for bloaty since the warning doesn't affect the coverage data as far as I could see. Or should I treat bloaty as a problematic benchmark and not generating anything from it?

**Comment by jonathanmetzman** [comment]
>Why not just use gcr.io/fuzzbench/base-builder image which has all these tools.

I don't see what benefit this provides but it does add a dependency of dispatcher-image on base-builder which is bad as it complicates things. I therefore think this shouldn't be changed

File: .gitignore [hunk 7]
```diff
@@ -36,3 +36,4 @@ docs/vendor/
 
 # Auto-generated build files.
 docker/generated.mk
+
```
Comment (jonathanmetzman) [inline .gitignore:39]: Don't make unnecessary changes. Remove this addition please.

File: experiment/measurer.py [hunk 8]
```diff
@@ -548,22 +559,26 @@ def measure_snapshot_coverage(fuzzer: str, benchmark: str, trial_num: int,
...
     # Record the new corpus files.
     snapshot_measurer.update_measured_files()
 
     # Archive crashes directory.
     snapshot_measurer.archive_crashes(cycle)
-
-    measuring_time = round(time.time() - measuring_start_time, 2)
-    snapshot_logger.info('Measured cycle: %d in %d seconds.', cycle,
-                         measuring_time)
+    measuring_time = round(time.time() - measuring_start_time, 4)
```
Comment (jonathanmetzman) [inline experiment/measurer.py:579]: Why is this needed at all if I already measure the time on 521?

File: experiment/run_coverage.py [hunk 9]
```diff
@@ -12,7 +12,7 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-"""Module for running a sancov instrumented binary on a corpus."""
+"""Module for running a clang-source-base instrumented binary on a corpus."""
```
Comment (jonathanmetzman) [inline experiment/run_coverage.py:15]: clang source-based coverage

**Comment by jonathanmetzman** [comment]
WHy does this PR update the commit in third_party/oss-fuzz? It shouldn't, please undo this change.

File: experiment/test_data/json_summary.txt [hunk 10]
```diff
@@ -0,0 +1 @@
+{"version":"2.0.0","type":"llvm.coverage.json.export","data":[{"files":[{"filename":"/home/ruikangshi/test/fuzzing/tutorial/libFuzzer/fuzz_me.cc","summary":{"lines":{"count":11,"covered":11,"percent":100},"functions":{"count":2,"covered":2,"percent":100},"instantiations":{"count":2,"covered":2,"percent":100},"regions":{"count":10,"covered":7,"notcovered":3,"percent":70}}}],"totals":{"lines":{"count":11,"covered":11,"percent":100},"functions":{"count":2,"covered":2,"percent":100},"instantiations":{"count":2,"covered":2,"percent":100},"regions":{"count":10,"covered":7,"notcovered":3,"percent":70}}}]}
```
Comment (jonathanmetzman) [inline experiment/test_data/json_summary.txt:1]: Why end in .txt if this is a json file?

File: experiment/test_measurer.py [hunk 11]
```diff
@@ -55,30 +56,84 @@ def db_experiment(experiment_config, db):
     yield
 
 
-@pytest.mark.parametrize('new_pcs', [['0x1', '0x2'], []])
-def test_merge_new_pcs(new_pcs, fs, experiment):
-    """Tests that merge_new_pcs merges new PCs, and updates the covered-pcs
-    file."""
+def test_get_current_coverage(fs, experiment):
+    """Tests that get_current_coverage read the correct data from json file"""
```
Comment (jonathanmetzman) [inline experiment/test_measurer.py:60]: "reads" and you need to end this with a period.

**Comment by inferno-chromium** [comment]
> > Why not just use gcr.io/fuzzbench/base-builder image which has all these tools.
> 
> I don't see what benefit this provides but it does add a dependency of dispatcher-image on base-builder which is bad as it complicates things. I therefore think this shouldn't be changed

That is fine, not a big deal, i thought we should minimize images inheriting from oss-fuzz-base, but that is fine as-is. @Shadoom7 - feel free to revert that part of change.

File: experiment/measurer.py (hunk 8 above)
Comment (Shadoom7) [reply experiment/measurer.py:579]: We need to calculate the time spent to measure one cycle. Line 521 is the start time and here we get the end time and then do a subtraction.
